### PR TITLE
udevadm verify

### PR DESCRIPTION
**Pull request by ldv-alt**
Implement a udev rules syntax checker in the form of
`udevadm verify [OPTIONS] FILE...` command that is based on
`udev_rules_parse_file` interface and would apply further checks
on top of it in the future.

Resolves: #26606

### CODE REVIEW

File: shell-completion/bash/udevadm [hunk 1]
```diff
@@ -247,6 +248,24 @@ _udevadm() {
             fi
             ;;
 
+        'verify')
```
Comment (yuwata) [inline shell-completion/bash/udevadm:251]: Maybe better to suggest `early` or `never`? See the line `'test'` in the above.

File: src/udev/udevadm-verify.c [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        r = udev_rules_parse_file(rules, fname);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse rules file %s: %m", fname);
+
+        unsigned int issues = udev_check_current_rule_file(rules);
+        unsigned int mask = (1U << LOG_ERR) | (1U << LOG_WARNING);
+        if (issues & mask)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "%s: udev rules check failed", fname);
+
+        return EXIT_SUCCESS;
```
Comment (yuwata) [inline src/udev/udevadm-verify.c]: Please return 0 on success.

File: src/udev/udevadm-verify.c [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+static int verify_rules(UdevRules *rules, unsigned int num_args, char **args) {
+        int rv = EXIT_SUCCESS;
+
+        for (unsigned int i = 0; i < num_args; ++i) {
+                int r = verify_rules_file(rules, args[i]);
```
Comment (yuwata) [inline src/udev/udevadm-verify.c]: Please always declare the special variable `int r;` at the beginning of the function, and return 0 on success:
```c
static int verify_rules(UdevRules *rules, unsigned int num_args, char **args) {
        int r, rv = 0;

        for (...) {
                r = verify_rules_file(...);
                if (r < 0 && rv >= 0)
                        rv = r;
        }

        return rv;
}
```

**Review by yuwata — CHANGES_REQUESTED** [review]
Sounds super nice!

Reviewed first half commits.

**Comment by yuwata** [comment]
```
2023-03-07T01:54:34.0327529Z  390/1433 systemd:dist-check / udev-rules-check                                                   FAIL              0.14s   exit status 1
2023-03-07T01:54:34.0332324Z >>> LD_LIBRARY_PATH=/home/runner/work/systemd/systemd/build/src/shared MALLOC_PERTURB_=230 /home/runner/work/systemd/systemd/build/udevadm verify /home/runner/work/systemd/systemd/build/../rules.d/60-autosuspend.rules /home/runner/work/systemd/systemd/build/../rules.d/60-block.rules /home/runner/work/systemd/systemd/build/../rules.d/60-cdrom_id.rules /home/runner/work/systemd/systemd/build/../rules.d/60-drm.rules /home/runner/work/systemd/systemd/build/../rules.d/60-evdev.rules /home/runner/work/systemd/systemd/build/../rules.d/60-fido-id.rules /home/runner/work/systemd/systemd/build/../rules.d/60-infiniband.rules /home/runner/work/systemd/systemd/build/../rules.d/60-input-id.rules /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-alsa.rules /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-input.rules /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage-tape.rules /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-v4l.rules /home/runner/work/systemd/systemd/build/../rules.d/60-sensor.rules /home/runner/work/systemd/systemd/build/../rules.d/60-serial.rules /home/runner/work/systemd/systemd/build/../rules.d/70-camera.rules /home/runner/work/systemd/systemd/build/../rules.d/70-joystick.rules /home/runner/work/systemd/systemd/build/../rules.d/70-mouse.rules /home/runner/work/systemd/systemd/build/../rules.d/70-touchpad.rules /home/runner/work/systemd/systemd/build/../rules.d/75-net-description.rules /home/runner/work/systemd/systemd/build/../rules.d/75-probe_mtd.rules /home/runner/work/systemd/systemd/build/../rules.d/78-sound-card.rules /home/runner/work/systemd/systemd/build/../rules.d/80-net-setup-link.rules /home/runner/work/systemd/systemd/build/../rules.d/81-net-dhcp.rules /home/runner/work/systemd/systemd/build/../rules.d/70-memory.rules /home/runner/work/systemd/systemd/build/../rules.d/70-power-switch.rules rules.d/50-udev-default.rules rules.d/64-btrfs.rules rules.d/99-systemd.rules rules.d/71-seat.rules rules.d/73-seat-late.rules rules.d/90-vconsole.rules
2023-03-07T01:54:34.0368396Z ――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
2023-03-07T01:54:34.0369207Z stderr:
2023-03-07T01:54:34.0370171Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:114 Unknown builtin command: blkid --hint=session_offset=$env{ID_CDROM_MEDIA_SESSION_LAST_OFFSET}
2023-03-07T01:54:34.0371067Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:117 Unknown builtin command: blkid --noraid
2023-03-07T01:54:34.0371740Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:120 Unknown builtin command: blkid
2023-03-07T01:54:34.0372390Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules: udev rules check failed
2023-03-07T01:54:34.0372929Z ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```

File: src/udev/udev-rules.c [hunk 4]
```diff
@@ -1152,13 +1159,15 @@ static void rule_resolve_goto(UdevRuleFile *rule_file) {
                 if (!line->goto_line) {
                         log_error("%s:%u: GOTO=\"%s\" has no matching label, ignoring",
                                   rule_file->filename, line->line_number, line->goto_label);
+                        rule_file_mark_issue(rule_file, LOG_ERR);
 
                         SET_FLAG(line->type, LINE_HAS_GOTO, false);
                         line->goto_label = NULL;
 
                         if ((line->type & ~LINE_HAS_LABEL) == 0) {
                                 log_notice("%s:%u: The line takes no effect any more, dropping",
```
Comment (yuwata) [inline src/udev/udev-rules.c]: This may be converted to `log_warning()`.

File: src/udev/udev-rules.c [hunk 5]
```diff
@@ -1152,13 +1159,15 @@ static void rule_resolve_goto(UdevRuleFile *rule_file) {
                 if (!line->goto_line) {
                         log_error("%s:%u: GOTO=\"%s\" has no matching label, ignoring",
                                   rule_file->filename, line->line_number, line->goto_label);
+                        rule_file_mark_issue(rule_file, LOG_ERR);
```
Comment (yuwata) [inline src/udev/udev-rules.c]: Maybe, better to introduce `log_line_error()` and friends that takes `UdevRuleLine` object.

File: src/udev/udev-rules.c [hunk 6]
```diff
@@ -1287,6 +1287,34 @@ int udev_rules_parse_file(UdevRules *rules, const char *filename) {
         return 0;
 }
 
+static bool token_data_is_string(UdevRuleTokenType type) {
+        if (IN_SET(type, TK_A_ENV, TK_M_ENV, TK_M_CONST, TK_A_ATTR, TK_M_ATTR,
+                         TK_A_SYSCTL, TK_M_SYSCTL, TK_M_PARENTS_ATTR, TK_A_SECLABEL))
+                return true;
+        return false;
```
Comment (yuwata) [inline src/udev/udev-rules.c]:
```c
return IN_SET(type,
              TK_A_ENV, TK_M_ENV, TK_M_CONST, TK_A_ATTR, TK_M_ATTR,
              TK_A_SYSCTL, TK_M_SYSCTL, TK_M_PARENTS_ATTR, TK_A_SECLABEL);
```

File: src/udev/udev-rules.c [hunk 7]
```diff
@@ -1342,6 +1347,25 @@ static void udev_check_rule_line(UdevRuleFile *rule_file, const UdevRuleLine *li
                         break;
                 }
         }
+
+        /* check for conflicting tokens */
+        LIST_FOREACH(tokens, token, line->tokens) {
```
Comment (yuwata) [inline src/udev/udev-rules.c]: Please merge the loop introduced in the above.

**Comment by yuwata** [comment]
CentOS CI failure:
```
388/1468 systemd:dist-check / udev-rules-check                                                   FAIL              0.76s   exit status 1
>>> MALLOC_PERTURB_=166 LD_LIBRARY_PATH=/systemd-meson-build/src/shared /systemd-meson-build/udevadm verify /systemd-meson-build/../build/rules.d/60-autosuspend.rules /systemd-meson-build/../build/rules.d/60-block.rules /systemd-meson-build/../build/rules.d/60-cdrom_id.rules /systemd-meson-build/../build/rules.d/60-drm.rules /systemd-meson-build/../build/rules.d/60-evdev.rules /systemd-meson-build/../build/rules.d/60-fido-id.rules /systemd-meson-build/../build/rules.d/60-infiniband.rules /systemd-meson-build/../build/rules.d/60-input-id.rules /systemd-meson-build/../build/rules.d/60-persistent-alsa.rules /systemd-meson-build/../build/rules.d/60-persistent-input.rules /systemd-meson-build/../build/rules.d/60-persistent-storage.rules /systemd-meson-build/../build/rules.d/60-persistent-storage-tape.rules /systemd-meson-build/../build/rules.d/60-persistent-v4l.rules /systemd-meson-build/../build/rules.d/60-sensor.rules /systemd-meson-build/../build/rules.d/60-serial.rules /systemd-meson-build/../build/rules.d/70-camera.rules /systemd-meson-build/../build/rules.d/70-joystick.rules /systemd-meson-build/../build/rules.d/70-mouse.rules /systemd-meson-build/../build/rules.d/70-touchpad.rules /systemd-meson-build/../build/rules.d/75-net-description.rules /systemd-meson-build/../build/rules.d/75-probe_mtd.rules /systemd-meson-build/../build/rules.d/78-sound-card.rules /systemd-meson-build/../build/rules.d/80-net-setup-link.rules /systemd-meson-build/../build/rules.d/81-net-dhcp.rules /systemd-meson-build/../build/rules.d/80-drivers.rules /systemd-meson-build/../build/rules.d/70-memory.rules /systemd-meson-build/../build/rules.d/70-power-switch.rules rules.d/50-udev-default.rules rules.d/64-btrfs.rules rules.d/99-systemd.rules rules.d/70-uaccess.rules rules.d/71-seat.rules rules.d/73-seat-late.rules rules.d/90-vconsole.rules
â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€• âœ€  â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•
stderr:
rules.d/50-udev-default.rules:29 Unknown group 'dialout', ignoring
rules.d/50-udev-default.rules:75 Unknown group 'cdrom', ignoring
rules.d/50-udev-default.rules:76 Unknown group 'cdrom', ignoring
rules.d/50-udev-default.rules:77 Unknown group 'cdrom', ignoring
rules.d/50-udev-default.rules:78 Unknown group 'cdrom', ignoring
rules.d/50-udev-default.rules:79 Unknown group 'cdrom', ignoring
rules.d/50-udev-default.rules:81 Unknown group 'tape', ignoring
rules.d/50-udev-default.rules: udev rules check failed
â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•
```
Maybe, better to use `-N never` in the test.

**Comment by yuwata** [comment]
Please add basic tests in TEST-17-UDEV to verify that the command actually detect something spurious in rules file.

**Comment by ldv-alt** [comment]
> rules.d/50-udev-default.rules:29 Unknown group 'dialout', ignoring
> rules.d/50-udev-default.rules:75 Unknown group 'cdrom', ignoring
> rules.d/50-udev-default.rules:76 Unknown group 'cdrom', ignoring
> rules.d/50-udev-default.rules:77 Unknown group 'cdrom', ignoring
> rules.d/50-udev-default.rules:78 Unknown group 'cdrom', ignoring
> rules.d/50-udev-default.rules:79 Unknown group 'cdrom', ignoring
> rules.d/50-udev-default.rules:81 Unknown group 'tape', ignoring
> rules.d/50-udev-default.rules: udev rules check failed
> 
> Maybe, better to use `-N never` in the test.

I agree, the test results clearly show that.

**Comment by ldv-alt** [comment]
> When built without blkid, then udev-builtin-blkid is not built, then the verifier warn about the unknown builtin.
> 2023-03-07T01:54:34.0370171Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:114 Unknown builtin command: blkid --hint=session_offset=$env{ID_CDROM_MEDIA_SESSION_LAST_OFFSET}
> 2023-03-07T01:54:34.0371067Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:117 Unknown builtin command: blkid --noraid
> 2023-03-07T01:54:34.0371740Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:120 Unknown builtin command: blkid
> 2023-03-07T01:54:34.0372390Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules: udev rules check failed

Interesting, I didn't expect _that_. Does it mean that 60-persistent-storage.rules should be parametrized with something like `{% if HAVE_BLKID %}`?

File: src/udev/udevadm-verify.c [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                switch (c) {
+                case 'h':
+                        return help();
+                case 'V':
+                        return print_version();
+                case 'N':
+                        arg_resolve_name_timing = resolve_name_timing_from_string(optarg);
+                        if (arg_resolve_name_timing < 0)
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "--resolve-names= must be early or never");
```
Comment (keszybz) [inline src/udev/udevadm-verify.c]: Maybe add quotes around early or never? The sentence is hard to parse otherwise.

File: src/udev/udevadm-verify.c [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "hVN:", options, NULL)) >= 0)
+                switch (c) {
+                case 'h':
+                        return help();
+                case 'V':
+                        return print_version();
+                case 'N':
+                        arg_resolve_name_timing = resolve_name_timing_from_string(optarg);
```
Comment (keszybz) [inline src/udev/udevadm-verify.c:60]: resolve_name_timing_from_string shoudl return a real errno value (it probably already does if it's using the standard machinery) and SYNTHETIC_ERRNO should not be used here.

File: src/udev/udevadm-verify.c [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        rules = udev_rules_new(arg_resolve_name_timing);
+        if (!rules)
+                return -ENOMEM;
+
+        return verify_rules(rules, argc - optind, argv + optind);
```
Comment (keszybz) [inline src/udev/udevadm-verify.c]: Please use `strv_skip()`.

File: src/udev/udev-rules.c (hunk 4 above)
Comment (keszybz) [reply src/udev/udev-rules.c]: I think `notice` is enough. I don't think we want to log at warning level for something that is not really an issue.

File: src/udev/udev-rules.c (hunk 6 above)
Comment (keszybz) [reply src/udev/udev-rules.c]: Maybe even make it one-per-line?

**Review by keszybz — COMMENTED** [review]
I think `udevadm verify` with no arguments should simply load all rules files from the system, like udevd does, and verify them.

**Comment by yuwata** [comment]
> > When built without blkid, then udev-builtin-blkid is not built, then the verifier warn about the unknown builtin.
> > 2023-03-07T01:54:34.0370171Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:114 Unknown builtin command: blkid --hint=session_offset=$env{ID_CDROM_MEDIA_SESSION_LAST_OFFSET}
> > 2023-03-07T01:54:34.0371067Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:117 Unknown builtin command: blkid --noraid
> > 2023-03-07T01:54:34.0371740Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules:120 Unknown builtin command: blkid
> > 2023-03-07T01:54:34.0372390Z /home/runner/work/systemd/systemd/build/../rules.d/60-persistent-storage.rules: udev rules check failed
> 
> Interesting, I didn't expect _that_. Does it mean that 60-persistent-storage.rules should be conditionalized with something like `{% if HAVE_BLKID %}`?

Yeah, I think so.

File: src/udev/udev-rules.c (hunk 6 above)
Comment (yuwata) [reply src/udev/udev-rules.c]: Yeah, that should be better.

File: src/udev/udev-rules.c (hunk 4 above)
Comment (yuwata) [reply src/udev/udev-rules.c]: IIUC, the log is shown when some tokens in a non-empty line are dropped and now the line becomes meaningless.
That should be not intentional in most cases, and the line should have some typo or so.
So, `udevadm verify` should detect such lines. Of course, the command reuses the parser, hence it logs about that, but succeeds unfortunately.

I do not think we should keep the log level.

Comment (ldv-alt) [reply src/udev/udev-rules.c]: Since there is a `log_error()` for the same line above, I don't yet see how upgrading that `log_notice()` to `log_warning()` could be beneficial to the verifier: this wouldn't affect the exit status, and the message is logged anyway.

**Comment by ldv-alt** [comment]
I decided to leave some of new checks that need more work to another PR.

**Comment by ldv-alt** [comment]
> I think `udevadm verify` with no arguments should simply load all rules files from the system, like udevd does, and verify them.

I'm not sure how to test this properly.

**Comment by ldv-alt** [comment]
> I think `udevadm verify` with no arguments should simply load all rules files from the system, like udevd does, and verify them.

What do you think if `udevadm verify` would have `--prefix` or `--root` option specifying the filesystem prefix where to load all udev rules when invoked without arguments?  This might be convenient for use e.g. in brp scripts that check %buildroot.

File: src/udev/udev-rules.c (hunk 4 above)
Comment (yuwata) [reply src/udev/udev-rules.c]: > Since there is a `log_error()` for the same line above, I don't yet see how upgrading that `log_notice()` to `log_warning()` could be beneficial to the verifier: this wouldn't affect the exit status, and the message is logged anyway.

Indeed. I overlooked the first error log.

**Comment by ldv-alt** [comment]
By the way, I spotted a typo in the commit message of the commit that introduces the new command, so I'll need to amend it.

File: src/udev/udev-rules.c [hunk 11]
```diff
@@ -254,6 +257,48 @@ struct UdevRules {
                               "Invalid value \"%s\" for %s (char %zu: %s), ignoring.", \
                               value, key, offset, hint)
 
+#define log_line_full_errno_zerook(rule_file, line_nr, level, error, fmt, ...) \
```
Comment (yuwata) [inline src/udev/udev-rules.c:261]: Why not to take `UdevRuleLine*`, and use `UdevRuleLine.rule_file` and `UdevRuleLine.line_number`?

File: src/udev/udev-rules.c [hunk 12]
```diff
@@ -1321,6 +1321,16 @@ int udev_rules_parse_file(UdevRules *rules, const char *filename) {
         return 0;
 }
 
+unsigned int udev_check_current_rule_file(UdevRules *rules) {
```
Comment (yuwata) [inline src/udev/udev-rules.c]: Please drop `int`, that is `unsigned int` -> `unsigned`.

File: src/udev/udev-rules.h [hunk 13]
```diff
@@ -18,6 +18,7 @@ typedef enum {
 } UdevRuleEscapeType;
 
 int udev_rules_parse_file(UdevRules *rules, const char *filename);
+unsigned int udev_check_current_rule_file(UdevRules *rules);
```
Comment (yuwata) [inline src/udev/udev-rules.h]: `unsigned int` -> `unsigned`

File: src/udev/udevadm-verify.c [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+                switch (c) {
+                case 'h':
+                        return help();
+                case 'V':
+                        return print_version();
+                case 'N':
+                        arg_resolve_name_timing = resolve_name_timing_from_string(optarg);
+                        if (arg_resolve_name_timing < 0)
+                                return log_error_errno(arg_resolve_name_timing,
+                                                       "--resolve-names= must be \"early\" or \"never\"");
```
Comment (yuwata) [inline src/udev/udevadm-verify.c]: `must be` -> `must take` ?

File: src/udev/udevadm-verify.c [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+        unsigned int issues = udev_check_current_rule_file(rules);
+        unsigned int mask = (1U << LOG_ERR) | (1U << LOG_WARNING);
+        if (issues & mask)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "%s: udev rules check failed", fname);
+
+        return 0;
+}
+
+static int verify_rules(UdevRules *rules, char *argv[]) {
```
Comment (yuwata) [inline src/udev/udevadm-verify.c]: Please rename the argument, as it is not argv anymore: `char *argv[]` -> `char **files`

File: src/udev/udevadm-verify.c [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+static int verify_rules_file(UdevRules *rules, const char *fname) {
+        int r;
+
+        r = udev_rules_parse_file(rules, fname);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse rules file %s: %m", fname);
+
+        unsigned int issues = udev_check_current_rule_file(rules);
+        unsigned int mask = (1U << LOG_ERR) | (1U << LOG_WARNING);
```
Comment (yuwata) [inline src/udev/udevadm-verify.c]: Please use `unsigned`, instead of `unsigned int`.

File: src/udev/udev-rules.c [hunk 17]
```diff
@@ -1321,13 +1324,25 @@ int udev_rules_parse_file(UdevRules *rules, const char *filename) {
         return 0;
 }
 
+static void udev_check_rule_line(UdevRuleFile *rule_file, const UdevRuleLine *line) {
+        /* check for unused labels */
+        if (FLAGS_SET(line->type, LINE_HAS_LABEL) &&
+            !FLAGS_SET(line->type, LINE_IS_REFERENCED)) {
```
Comment (yuwata) [inline src/udev/udev-rules.c]: Please drop unnecessary `{ }`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks pretty good to me. Several minor comments.

**Comment by yuwata** [comment]
> > I think `udevadm verify` with no arguments should simply load all rules files from the system, like udevd does, and verify them.
> 
> What do you think if `udevadm verify` would have `--prefix` or `--root` option specifying the filesystem prefix where to load all udev rules when invoked without arguments? This might be convenient for use e.g. in brp scripts that check %buildroot.

That sounds nice.

File: src/udev/udev-rules.c (hunk 11 above)
Comment (ldv-alt) [reply src/udev/udev-rules.c:261]: Makes sense.

File: src/udev/udev-rules.c (hunk 12 above)
Comment (ldv-alt) [reply src/udev/udev-rules.c]: Ouch, that's unusual.  I haven't seen anything about it in the docs.
Is this a policy?  If yes, I suppose I can document it somewhere in `docs/CODING_STYLE.md`.

File: src/udev/udevadm-verify.c (hunk 14 above)
Comment (ldv-alt) [reply src/udev/udevadm-verify.c]: Agreed.  By the way, the message was copied from another file, worth changing it as well.

File: src/udev/udevadm-verify.c (hunk 15 above)
Comment (ldv-alt) [reply src/udev/udevadm-verify.c]: ok

File: src/udev/udev-rules.c (hunk 11 above)
Comment (ldv-alt) [reply src/udev/udev-rules.c:261]: Yes, it would be nice to have just one argument instead of two, but there is one place in the code where `Line is too long` error is logged but there is no associated `UdevRuleLine` object because `rule_add_line` has not been called for that invalid line.

What would you suggest?

File: src/udev/udev-rules.c (hunk 12 above)
Comment (ldv-alt) [reply src/udev/udev-rules.c]: Posted a patch at #26718.

Comment (yuwata) [reply src/udev/udev-rules.c]: Thank you!

File: src/udev/udev-rules.c (hunk 11 above)
Comment (yuwata) [reply src/udev/udev-rules.c:261]: Indeed. Let's cleanup the macros later, if we find a nicer way.

**Review by yuwata — APPROVED** [review]
